File: backend/schemas/report.py
from datetime import datetime
from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional

class ReportBase(BaseModel):
    title: str
    summary: Optional[str] = None
    report_type: str

class ReportCreate(ReportBase):
    pass

class GenerateReportRequest(BaseModel):
    title: str
    report_type: str

class ReportTypeInfo(BaseModel):
    key: str
    name: str
    description: str

class ReportTypeResponse(BaseModel):
    types: List[ReportTypeInfo]

class ReportSummaryResponse(BaseModel):
    id: int
    report_id: int
    title: str
    report_type: str
    generated_at: datetime
    executive_summary: Optional[str] = None

    class Config:
        from_attributes = True

class ReportListResponse(BaseModel):
    reports: List[ReportSummaryResponse]

class CrimeCategoryBreakdown(BaseModel):
    category: str
    count: int

class CrimeOverview(BaseModel):
    total_crimes: int
    top_categories: List[CrimeCategoryBreakdown]
    trend_direction: str

class HighRiskLocation(BaseModel):
    district: str
    crime_count: int
    risk_level: str

class PredictiveInsights(BaseModel):
    high_risk_locations: List[HighRiskLocation]
    hotspot_count: int
    risk_score_summary: Dict[str, Any]

class NetworkKeyEntity(BaseModel):
    id: str
    type: str
    label: str
    score: float

class NetworkInsights(BaseModel):
    total_networks: int
    largest_network_size: int
    key_entities: List[NetworkKeyEntity]

class ReportResponse(BaseModel):
    report_id: int
    report_type: str
    title: str
    generated_at: datetime
    executive_summary: str
    crime_overview: CrimeOverview
    predictive_insights: PredictiveInsights
    network_insights: NetworkInsights
    recommendations: List[Dict[str, Any]]
    alerts: List[Dict[str, Any]]
    dataset_name: Optional[str] = None
    model_version: Optional[str] = None
    prediction_accuracy: Optional[float] = None

    class Config:
        from_attributes = True

File: frontend/features/reports/types/report.ts
export interface ReportType {
  key: string;
  name: string;
  description: string;
}

export interface ReportSummary {
  id: number;
  report_id: number;
  title: string;
  report_type: string;
  generated_at: string;
  executive_summary?: string;
}

export interface CrimeCategoryBreakdown {
  category: string;
  count: number;
}

export interface CrimeOverview {
  total_crimes: number;
  top_categories: CrimeCategoryBreakdown[];
  trend_direction: string;
}

export interface HighRiskLocation {
  district: string;
  crime_count: number;
  risk_level: string;
}

export interface PredictiveInsights {
  high_risk_locations: HighRiskLocation[];
  hotspot_count: number;
  risk_score_summary: {
    average_criminal_risk: number;
    high_risk_criminals_count: number;
    total_criminals_tracked: number;
  };
}

export interface NetworkKeyEntity {
  id: string;
  type: string;
  label: string;
  score: number;
}

export interface NetworkInsights {
  total_networks: number;
  largest_network_size: number;
  key_entities: NetworkKeyEntity[];
}

export interface Recommendation {
  id: number;
  priority: string;
  recommendation_text: string;
  reason: string;
  status: string;
}

export interface Alert {
  id: number;
  title: string;
  description: string;
  severity: string;
  source: string;
  status: string;
}

export interface Report {
  report_id: number;
  report_type: string;
  title: string;
  generated_at: string;
  executive_summary: string;
  crime_overview: CrimeOverview;
  predictive_insights: PredictiveInsights;
  network_insights: NetworkInsights;
  recommendations: Recommendation[];
  alerts: Alert[];
  dataset_name?: string;
  model_version?: string;
  prediction_accuracy?: number;
}
